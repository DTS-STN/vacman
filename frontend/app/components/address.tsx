import type { ComponentProps, JSX } from 'react';

import { formatAddress } from '~/utils/string-utils';
import { cn } from '~/utils/tailwind-utils';

type AddressProps = ComponentProps<'address'> & {
  address: Parameters<typeof formatAddress>[0];
  format?: Parameters<typeof formatAddress>[1];
};

export function Address({ address, format, className, ...rest }: AddressProps): JSX.Element {
  return (
    <address className={cn('whitespace-pre-wrap not-italic', className)} data-testid="address-id" {...rest}>
      {formatAddress(address, format)}
    </address>
  );
}
