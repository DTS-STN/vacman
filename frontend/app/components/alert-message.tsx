import type { Ref } from 'react';

import type { IconDefinition } from '@fortawesome/free-solid-svg-icons';
import { faTriangleExclamation, faCircleExclamation, faCircleCheck } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface AlertMessageProps {
  message: string;
  type: 'success' | 'info' | 'error';
  ref?: Ref<HTMLDivElement>;
}

const styles: Record<NonNullable<AlertMessageProps['type']>, string> = {
  success: `border-green-600 bg-green-100`,
  error: `border-red-800 bg-red-100`,
  info: `border-sky-700 bg-sky-100`,
};

const icons: Record<NonNullable<AlertMessageProps['type']>, IconDefinition> = {
  success: faCircleCheck,
  error: faTriangleExclamation,
  info: faCircleExclamation,
};

export function AlertMessage({ ref, message, type = 'info' }: AlertMessageProps) {
  return (
    <div ref={ref} className={`${styles[type]} flex w-full items-center border-l-4 p-2`}>
      <FontAwesomeIcon
        icon={icons[type]}
        className={`${type === 'success' ? 'text-green-600' : type === 'error' ? 'text-red-800' : 'text-sky-700'} mr-2`}
      />
      {message}
    </div>
  );
}
