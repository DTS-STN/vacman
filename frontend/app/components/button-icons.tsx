import type { FontAwesomeIconProps } from '@fortawesome/react-fontawesome';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { cn } from '~/utils/tailwind-utils';

export function ButtonStartIcon({ className, ...restProps }: FontAwesomeIconProps) {
  return <FontAwesomeIcon fixedWidth className={cn('me-1', className)} {...restProps} />;
}

export function ButtonEndIcon({ className, ...restProps }: FontAwesomeIconProps) {
  return <FontAwesomeIcon fixedWidth className={cn('ms-1', className)} {...restProps} />;
}
