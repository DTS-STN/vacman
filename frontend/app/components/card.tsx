import type { ComponentProps, JSX } from 'react';

import type { FlipProp, IconProp } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Slot } from '@radix-ui/react-slot';

import { cn } from '~/utils/tailwind-utils';

type CardProps = ComponentProps<'div'> & {
  /**
   * Use the asChild prop to change the default rendered element for the one passed
   * as a child, merging their props and behavior. Useful for rendering links.
   */
  asChild?: boolean;
};

/**
 * Card component renders a card with optional child rendering.
 * It supports various customization options.
 */
export function Card({ className, asChild, ...props }: CardProps): JSX.Element {
  const Component = asChild ? Slot : 'div';

  return (
    <Component
      className={cn(
        'group block rounded-xs border bg-white shadow-xs [&:is(a)]:hover:bg-gray-50 [&:is(a)]:hover:shadow-md',
        className,
      )}
      {...props}
    />
  );
}

/**
 * CardHeader component renders the header section of a card.
 */
export function CardHeader({ className, ...props }: ComponentProps<'div'>) {
  return <div className={cn('flex flex-col space-y-1.5 p-6', className)} {...props} />;
}

type CardTitleProps = ComponentProps<'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'> & {
  /**
   * Use the asChild prop to change the default rendered element for the one passed as a child, merging their props and behavior.
   */
  asChild?: boolean;
};

/**
 * CardTitle component renders the title of a card.
 * It supports optional child rendering.
 */
export function CardTitle({ asChild, className, ...props }: CardTitleProps) {
  const Component = asChild ? Slot : 'h3';

  return (
    <Component
      className={cn(
        'leading-none font-semibold group-[&:is(a)]:text-slate-700 group-[&:is(a)]:underline group-[&:is(a):focus]:text-blue-700 group-[&:is(a):hover]:text-blue-700',
        className,
      )}
      {...props}
    />
  );
}

/**
 * CardDescription component renders the description of a card.
 */
export function CardDescription({ className, ...props }: ComponentProps<'div'>) {
  return <div className={cn('text-sm text-black/60', className)} {...props} />;
}

/**
 * CardContent component renders the content section of a card.
 */
export function CardContent({ className, ...props }: ComponentProps<'div'>) {
  return <div className={cn('p-6 pt-0', className)} {...props} />;
}

/**
 * CardFooter component renders the footer section of a card.
 */
export function CardFooter({ className, ...props }: ComponentProps<'div'>) {
  return <div className={cn('flex items-center p-6 pt-0', className)} {...props} />;
}

/**
 * CardImage component renders an image in a card.
 */
export function CardImage({ className, ...props }: ComponentProps<'img'>) {
  return (
    <img
      alt=""
      className={cn('relative aspect-video h-40 w-full object-cover sm:h-48', className)}
      loading="lazy"
      draggable="false"
      {...props}
    />
  );
}

/**
 * CardTag component renders a tag in a card.
 */
export function CardTag({ className, ...props }: ComponentProps<'span'>) {
  return <span className={cn('inline-block bg-cyan-700 px-2 py-1 text-xs text-white', className)} {...props} />;
}

type CardIconProps = OmitStrict<ComponentProps<'div'>, 'children'> & {
  /**
   * FontAwesome icon to be displayed.
   */
  icon: IconProp;
  /**
   * Additional class names for styling the icon.
   */
  iconClassName?: IconProp;
  /**
   * Flip property for the icon.
   */
  iconFlip?: FlipProp;
};

/**
 * CardIcon component renders an icon in a card.
 */
export function CardIcon({ className, icon, iconClassName, iconFlip, ...props }: CardIconProps) {
  return (
    <div className="flex size-10 shrink-0 items-center justify-center rounded-full bg-slate-700 sm:size-12" {...props}>
      <FontAwesomeIcon icon={icon} flip={iconFlip} className={cn('size-4 text-white sm:size-5', iconClassName)} />
    </div>
  );
}
