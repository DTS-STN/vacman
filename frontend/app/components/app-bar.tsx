import type { JSX } from 'react';

import { faChevronDown, faRightFromBracket, faUser } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { DropdownMenu, DropdownMenuContent, DropdownMenuLabel, DropdownMenuTrigger } from '~/components/dropdown-menu';
import { LanguageSwitcher } from '~/components/language-switcher';
import { Menu, MenuItem } from '~/components/menu';
import { cn } from '~/utils/tailwind-utils';

type AppBarProps = {
  children: React.ReactNode;
  name?: string;
  profileItems?: React.ReactNode;
};

export function AppBar({ children, name, profileItems }: AppBarProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);

  return (
    <div className="bg-slate-700">
      <div className="align-center container mx-auto flex flex-wrap justify-between">
        <div className="align-center flex">
          <Menu>{children}</Menu>
        </div>
        <div className="flex items-center space-x-4 text-right">
          {name && <UserButton name={name}>{profileItems}</UserButton>}
          <LanguageSwitcher>{t('gcweb:language-switcher.alt-lang')}</LanguageSwitcher>
        </div>
      </div>
    </div>
  );
}

type UserButtonProps = {
  className?: string;
  children?: React.ReactNode;
  name?: string;
};

function UserButton({ className, children, name }: UserButtonProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);

  return (
    <DropdownMenu>
      <DropdownMenuTrigger
        className={cn(
          'flex h-full flex-nowrap space-x-2 bg-slate-600 px-2 text-sm text-white hover:bg-slate-500 hover:underline focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-hidden aria-expanded:bg-slate-800 aria-expanded:text-white sm:space-x-4 sm:px-4',
          className,
        )}
      >
        <div className="text-md my-auto flex flex-nowrap items-center space-x-2 py-2">
          <div className="flex h-8 w-8 items-center justify-center rounded-full bg-white">
            <FontAwesomeIcon icon={faUser} className="size-5 text-slate-700" />
          </div>
          <span id="menu-label" className="text-md hidden py-2 font-bold sm:block">
            {name}
          </span>
        </div>
        <FontAwesomeIcon icon={faChevronDown} className="my-auto size-4" />
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end" className="w-64 bg-slate-700">
        <UserName name={name} />
        <MenuItem file="routes/protected/index.tsx">{t('gcweb:app.profile')}</MenuItem>
        {children}
        <MenuItem to="/auth/logout" className="text-md flex justify-between">
          {t('gcweb:app.logout')}
          <FontAwesomeIcon icon={faRightFromBracket} className="my-auto size-8" />
        </MenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  );
}

type UserNameProps = {
  name?: string;
};

function UserName({ name }: UserNameProps): JSX.Element {
  return (
    <>
      {name !== undefined && (
        <DropdownMenuLabel className="text-md flex items-center border-b-2 border-slate-600 px-3 py-2 text-gray-300 sm:hidden">
          <FontAwesomeIcon icon={faUser} className="mr-2 size-4" />
          {name}
        </DropdownMenuLabel>
      )}
    </>
  );
}
