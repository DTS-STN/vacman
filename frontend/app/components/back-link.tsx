import type { Params, Path } from 'react-router';

import { faAngleLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { InlineLink } from '~/components/links';
import type { I18nRouteFile } from '~/i18n-routes';
import { cn } from '~/utils/tailwind-utils';

interface BackLinkProps {
  file: I18nRouteFile;
  params?: Params;
  className?: string;
  hash?: Path['hash'];
  search?: Path['search'];
  translationKey?: string;
}

export function BackLink({ className, file, params, hash, search, translationKey = 'app:profile.back' }: BackLinkProps) {
  const { t } = useTranslation();
  const linkText = t(translationKey);
  return (
    <InlineLink
      className={cn('inline-flex items-center', className)}
      file={file}
      params={params}
      hash={hash}
      search={search}
      aria-label={linkText}
    >
      <FontAwesomeIcon icon={faAngleLeft} />
      {linkText}
    </InlineLink>
  );
}
