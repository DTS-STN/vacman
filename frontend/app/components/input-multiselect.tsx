import { useState, useRef, useEffect } from 'react';
import type { ComponentProps, ReactNode } from 'react';

import { faChevronDown, faChevronUp } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { InputCheckbox } from '~/components/input-checkbox';
import { InputError } from '~/components/input-error';
import { InputHelp } from '~/components/input-help';
import { InputLegend } from '~/components/input-legend';
import { cn } from '~/utils/tailwind-utils';

const triggerBaseClassName =
  'relative flex w-full cursor-pointer items-center justify-between rounded-lg border border-gray-500 bg-white px-3 py-2 text-left focus:border-blue-500 focus:outline-none focus:ring focus:ring-blue-500';
const triggerDisabledClassName =
  'disabled:bg-gray-100 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-70';
const triggerErrorClassName = 'border-red-500 focus:border-red-500 focus:ring-red-500';
const dropdownClassName =
  'absolute z-10 mt-1 max-h-60 w-full overflow-auto rounded-md bg-white py-1 text-base shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none';

export interface MultiSelectOption
  extends OmitStrict<
    ComponentProps<typeof InputCheckbox>,
    'aria-describedby' | 'aria-errormessage' | 'aria-invalid' | 'aria-required' | 'id' | 'name' | 'required'
  > {
  label: ReactNode;
  value: string;
}

export interface InputMultiSelectProps extends OmitStrict<ComponentProps<'div'>, 'id' | 'onChange'> {
  ariaDescribedbyId?: string;
  value: string[];
  onChange: (selectedValues: string[]) => void;
  options: MultiSelectOption[];
  id: string;
  legend: string;
  name: string;
  errorMessage?: string;
  helpMessage?: ReactNode;
  placeholder?: string;
  required?: boolean;
  disabled?: boolean;
}

export function InputMultiSelect(props: InputMultiSelectProps) {
  const {
    ariaDescribedbyId,
    errorMessage,
    helpMessage,
    id,
    legend,
    options,
    className,
    required,
    value,
    onChange,
    disabled,
    placeholder = 'Select options...',
    name,
    ...restDivProps
  } = props;

  const { t } = useTranslation(['gcweb']);
  const [isOpen, setIsOpen] = useState(false);
  const wrapperRef = useRef<HTMLDivElement>(null);

  const inputErrorId = `input-${id}-error`;
  const inputHelpMessageId = `input-${id}-help`;
  const inputLegendId = `input-${id}-legend`;
  const inputWrapperId = `input-${id}`;
  const dropdownId = `${id}-dropdown`;

  useEffect(() => {
    function handleClickOutside(event: MouseEvent) {
      if (wrapperRef.current && !wrapperRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    }
    function handleKeyDown(event: KeyboardEvent) {
      if (event.key === 'Escape') {
        setIsOpen(false);
      }
    }
    document.addEventListener('mousedown', handleClickOutside);
    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, []);

  const handleOptionClick = (optionValue: string) => {
    const newSelectedValues = value.includes(optionValue) ? value.filter((v) => v !== optionValue) : [...value, optionValue];
    onChange(newSelectedValues);
  };

  const handleOptionKeyDown = (e: React.KeyboardEvent, optionValue: string) => {
    if (e.key === 'Enter' || e.key === ' ') {
      e.preventDefault();
      handleOptionClick(optionValue);
    }
  };

  const getDisplayValue = () => {
    if (value.length === 0) {
      return <span className="text-gray-500">{placeholder}</span>;
    }
    if (value.length === 1) {
      const selectedOption = options.find((opt) => opt.value === value[0]);
      return selectedOption?.label ?? placeholder;
    }
    return t('gcweb:input.input-items-selected', { count: value.length });
  };

  const ariaDescribedbyIds =
    [
      !!helpMessage && inputHelpMessageId, //
      !!ariaDescribedbyId && ariaDescribedbyId,
    ]
      .filter(Boolean)
      .join(' ') || undefined;

  return (
    <div className="relative" ref={wrapperRef} {...restDivProps}>
      <fieldset id={inputWrapperId} aria-describedby={ariaDescribedbyId}>
        <InputLegend id={inputLegendId} required={required} className="mb-2">
          {legend}
        </InputLegend>
        {errorMessage && (
          <p className="mb-2">
            <InputError id={inputErrorId}>{errorMessage}</InputError>
          </p>
        )}
        {value.map((val) => (
          <input type="hidden" name={name} key={val} value={val} />
        ))}
        {helpMessage && (
          <InputHelp id={inputHelpMessageId} className="mt-2 mb-2">
            {helpMessage}
          </InputHelp>
        )}

        <div
          id={id}
          role="combobox"
          tabIndex={disabled ? -1 : 0}
          aria-required={required ? true : undefined}
          aria-disabled={disabled}
          aria-haspopup="listbox"
          aria-expanded={isOpen}
          aria-controls={dropdownId}
          aria-invalid={!!errorMessage}
          aria-errormessage={errorMessage ? inputErrorId : undefined}
          aria-labelledby={inputLegendId}
          aria-describedby={ariaDescribedbyIds}
          onClick={disabled ? undefined : () => setIsOpen(!isOpen)}
          onKeyDown={
            disabled
              ? undefined
              : (e) => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    e.preventDefault(); // Prevent space from scrolling
                    setIsOpen(!isOpen);
                  }
                }
          }
          className={cn(
            triggerBaseClassName,
            disabled && triggerDisabledClassName,
            errorMessage && triggerErrorClassName,
            className,
          )}
        >
          <span className="block truncate">{getDisplayValue()}</span>
          <span className="pointer-events-none inset-y-0 right-0 flex items-center pr-2">
            {isOpen ? (
              <FontAwesomeIcon icon={faChevronUp} className="my-auto size-3 text-gray-400" />
            ) : (
              <FontAwesomeIcon icon={faChevronDown} className="my-auto size-3 text-gray-400" />
            )}
          </span>
        </div>
        {isOpen && (
          <div id={dropdownId} role="listbox" aria-multiselectable="true" className={dropdownClassName}>
            {options.map((optionProps, index) => {
              const isSelected = value.includes(optionProps.value);
              const optionId = `${id}-multiselect-option-${index}`;

              return (
                <div
                  key={optionProps.value}
                  id={optionId}
                  role="option"
                  aria-selected={isSelected}
                  onClick={() => handleOptionClick(optionProps.value)}
                  tabIndex={0}
                  onKeyDown={(e) => handleOptionKeyDown(e, optionProps.value)}
                  className="relative cursor-pointer p-2 text-gray-900 select-none hover:bg-gray-100 focus:bg-gray-100 focus:outline-none"
                >
                  <InputCheckbox
                    id={`${optionId}-checkbox`}
                    name={`${name}-${optionProps.value}`}
                    checked={isSelected}
                    readOnly={true}
                    required={required}
                    hasError={!!errorMessage}
                    tabIndex={-1}
                    className="pointer-events-none"
                    {...optionProps}
                  >
                    {optionProps.label}
                  </InputCheckbox>
                </div>
              );
            })}
          </div>
        )}
      </fieldset>
    </div>
  );
}
