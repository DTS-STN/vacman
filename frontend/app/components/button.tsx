import type { ComponentProps } from 'react';

import { ButtonEndIcon, ButtonStartIcon } from '~/components/button-icons';
import { cn } from '~/utils/tailwind-utils';

type ButtonEndIconProps = ComponentProps<typeof ButtonEndIcon>;
type ButtonStartIconProps = ComponentProps<typeof ButtonStartIcon>;

const sizes = {
  xs: 'px-3 py-2 text-xs',
  sm: 'px-3 py-2 text-sm',
  base: 'px-5 py-2.5 text-sm',
  lg: 'px-5 py-3 text-base',
  xl: 'px-6 py-3.5 text-base',
} as const;

// prettier-ignore
const variants = {
  alternative: 'border-gray-200 bg-white text-gray-900 hover:bg-gray-100 hover:text-blue-700 focus:bg-gray-100 focus:text-blue-700',
  default: 'border-gray-300 bg-gray-200 text-slate-700 hover:bg-neutral-300 focus:bg-neutral-300',
  dark: 'border-gray-800 bg-gray-800 text-white hover:bg-gray-900 focus:bg-gray-900',
  green: 'border-green-700 bg-green-700 text-white hover:bg-green-800 focus:bg-green-800',
  primary: 'border-slate-700 bg-slate-700 text-white hover:bg-sky-800 focus:bg-sky-800',
  red: 'border-red-700 bg-red-700 text-white hover:bg-red-800 focus:bg-red-800',
  link: 'border-0 bg-white text-gray-900 underline hover:bg-gray-100 hover:text-blue-700 focus:bg-gray-100 focus:text-blue-700',
  ghost: 'border-0 hover:bg-neutral-100 hover:text-neutral-900',
  outline: "border border-neutral-200 bg-white hover:bg-neutral-100 hover:text-neutral-900 disabled:text-neutral-500",
} as const;

export interface ButtonProps extends ComponentProps<'button'> {
  endIcon?: ButtonEndIconProps['icon'];
  endIconProps?: OmitStrict<ButtonEndIconProps, 'icon'>;
  pill?: boolean;
  size?: keyof typeof sizes;
  startIcon?: ButtonStartIconProps['icon'];
  startIconProps?: OmitStrict<ButtonStartIconProps, 'icon'>;
  variant?: keyof typeof variants;
}

/**
 * Tailwind CSS Buttons from Flowbite
 * @see https://flowbite.com/docs/components/buttons/
 */
export function Button({
  children,
  className,
  endIcon,
  endIconProps,
  pill,
  size = 'base',
  startIcon,
  startIconProps,
  variant = 'default',
  ...props
}: ButtonProps) {
  const buttonClassName = cn(
    'font-lato inline-flex items-center justify-center rounded border align-middle outline-offset-4',
    sizes[size],
    variants[variant],
    'disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-70',
    pill && 'rounded-full',
    className,
  );
  return (
    <button className={buttonClassName} {...props}>
      {startIcon && <ButtonStartIcon {...(startIconProps ?? {})} icon={startIcon} />}
      {children}
      {endIcon && <ButtonEndIcon {...(endIconProps ?? {})} icon={endIcon} />}
    </button>
  );
}
