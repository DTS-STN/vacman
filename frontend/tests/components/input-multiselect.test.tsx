import type { ComponentProps } from 'react';

import { render, screen } from '@testing-library/react';
import { vi, describe, it, expect, beforeEach } from 'vitest';

import { InputMultiSelect } from '~/components/input-multiselect';
import type { InputMultiSelectProps } from '~/components/input-multiselect';

// --- 1. Mock all dependencies ---

// Mock FontAwesomeIcon to simplify testing

vi.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: (props: { icon: { iconName: string } }) => <i data-icon={props.icon.iconName} />,
}));

// Mock child components to isolate the component under test
vi.mock('~/components/input-label', () => ({
  InputLabel: (props: ComponentProps<'label'>) => <label {...props}>{props.children}</label>,
}));
vi.mock('~/components/input-error', () => ({
  InputError: (props: ComponentProps<'span'>) => <span {...props}>{props.children}</span>,
}));
vi.mock('~/components/input-help', () => ({
  InputHelp: (props: ComponentProps<'p'>) => <p {...props}>{props.children}</p>,
}));
vi.mock('~/components/input-checkbox', () => ({
  InputCheckbox: ({ children, ...props }: ComponentProps<'input'>) => (
    <div>
      <input type="checkbox" {...props} />
      <label>{children}</label>
    </div>
  ),
}));

// --- 2. Test Suite Setup ---

const MOCK_OPTIONS = [
  { label: 'React', value: 'react' },
  { label: 'Vue', value: 'vue' },
  { label: 'Svelte', value: 'svelte' },
];

describe('InputMultiSelect', () => {
  let onChange: ReturnType<typeof vi.fn>;
  let defaultProps: InputMultiSelectProps;

  beforeEach(() => {
    onChange = vi.fn();
    defaultProps = {
      id: 'framework',
      name: 'frameworks',
      label: 'Favorite Frameworks',
      options: MOCK_OPTIONS,
      value: [],
      onChange,
    };
  });

  const renderComponent = (props: Partial<InputMultiSelectProps> = {}) => {
    return render(<InputMultiSelect {...defaultProps} {...props} />);
  };

  // --- 3. Tests ---

  it('should render correctly with a placeholder', () => {
    renderComponent({ placeholder: 'Choose one...' });
    expect(screen.getByText('Favorite Frameworks')).toBeTruthy();
    expect(screen.getByText('Choose one...')).toBeTruthy();
    expect(screen.queryByRole('listbox')).toBeNull();
  });

  it('should toggle the dropdown on button click', async () => {
    const user = userEvent.setup();
    renderComponent();

    const triggerButton = screen.getByRole('button', { name: 'Favorite Frameworks' });
    expect(triggerButton.getAttribute('aria-expanded')).toBe('false');

    await user.click(triggerButton);
    expect(screen.getByRole('listbox')).toBeTruthy();
    expect(triggerButton.getAttribute('aria-expanded')).toBe('true');

    await user.click(triggerButton);
    expect(screen.queryByRole('listbox')).toBeNull();
  });

  it('should call onChange when selecting an option', async () => {
    const user = userEvent.setup();
    renderComponent();

    await user.click(screen.getByRole('button'));
    await user.click(screen.getByText('Svelte'));

    expect(onChange).toHaveBeenCalledWith(['svelte']);
  });

  it('should select an option with the Enter key', async () => {
    const user = userEvent.setup();
    renderComponent();

    await user.click(screen.getByRole('button'));
    const vueOption = screen.getByText('Vue').closest('div[role="option"]');

    expect(vueOption).toBeInstanceOf(HTMLElement);
    (vueOption as HTMLElement).focus();

    await user.keyboard('{Enter}');
    expect(onChange).toHaveBeenCalledWith(['vue']);
  });

  it('should render an error message and apply error styles', () => {
    renderComponent({ errorMessage: 'This field is required' });
    expect(screen.getByText('This field is required')).toBeTruthy();
    const button = screen.getByRole('button');
    expect(button.className).toContain('border-red-500');
  });

  it('should be disabled when the disabled prop is true', () => {
    renderComponent({ disabled: true });
    const button = screen.getByRole('button') as HTMLButtonElement;
    expect(button.disabled).toBe(true);
  });

  it('should render multiple hidden inputs for form submission', () => {
    const { container } = renderComponent({ value: ['react', 'svelte'] });

    const hiddenInputs = container.querySelectorAll('input[type="hidden"][name="frameworks"]');
    expect(hiddenInputs).toHaveLength(2);

    const values = Array.from(hiddenInputs).map((input) => (input as HTMLInputElement).value);
    expect(values).toEqual(['react', 'svelte']);
  });

  it('should match the initial snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
