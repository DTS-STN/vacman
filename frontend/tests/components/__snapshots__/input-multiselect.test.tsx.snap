// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputMultiSelect > should match the initial snapshot 1`] = `
<div>
  <div
    class="relative"
    id="input-framework"
  >
    <label
      class="mb-2"
      for="framework"
      id="input-framework-label"
    >
      Favorite Frameworks
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="input-framework-label"
      class="relative flex w-full cursor-pointer items-center justify-between rounded-lg border border-gray-500 bg-white px-3 py-2 text-left focus:border-blue-500 focus:outline-none focus:ring focus:ring-blue-500"
      id="framework"
      type="button"
    >
      <span
        class="block truncate"
      >
        <span
          class="text-gray-500"
        >
          Select options...
        </span>
      </span>
      <span
        class="pointer-events-none inset-y-0 right-0 flex items-center pr-2"
      >
        <i
          data-icon="chevron-down"
        />
      </span>
    </button>
  </div>
</div>
`;
